Fixes #26709 - perform HTTP HEAD on inst media

lzap opened the pull request:
This change makes Foreman to perform HTTP HEAD request once per request
to solve two problems:

* When installation media points to incorrect host or 404 path, Foreman
will now nicely report error. Compare to success action but failure on
the TFTP proxy which led to provisioning error.

* Foreman also reads last-modified/etag headers and puts that into the
unique TFTP filename suffix which helps to avoid corrupted TFTP files
after OS updates.

I am putting the helper method into operating system intentionally
because we want to cache (memorize) the value. This can't be done in
Media Provider, then the HTTP HEAD request would be done not just once
but 4 or more times during orchestration which is too much to my taste.

Once merged, I am going to do Katello PR to make use of the helper method
to do the same (hash from REPO ID + LAST-MODIFIED). This will fix it for Katello
too.

lzap committed:
Fixes #26709 - perform HTTP HEAD on inst media

ShimShtein commented on app/models/operatingsystem.rb:
```diff
@@ -313,6 +313,28 @@ def pxe_kernel_options(params)
     options
   end
 
+  def last_modified_id(media_provider)
```
The same operating system can be installed from multiple media sources (even before we start the "flavours" saga). I think this shouldn't be on operating system object. Can I suggest moving this method to medium provider's `unique_id` and caching the whole medium provider? As for Katello, from talking to them, it seems that `kickstart_repo.id` is unique and will change with each promotion.

lzap replied:
I will move it but that will create a problem of making HTTP HEAD call several times per request. I found a way how to memorize media provider by moving `pxe_files` from OS to host and with a slight change in renderer.

As for Katello, promotion is irrelevant and we need to do what I suggest anyway. When you sync new RHEL into Library, there is no promotion involved so the ID stays the same, yet there is a new PXE files pair which will likely break `wget`.

timogoebel commented on app/models/operatingsystem.rb:
```diff
@@ -313,6 +313,28 @@ def pxe_kernel_options(params)
+        begin
+          logger.debug "Performing HTTP HEAD #{url}"
+          uri = URI(url)
+          Net::HTTP.start(uri.host, uri.port) do |http|
```
Do we need to ask the smart- 
proxy to do this? The proxy currently downloads the file, does foreman core have access to it? I‘m thinking of firewalls etc.

timogoebel replied:
I just remembered there was an attempt so solve this in smart-proxy: https://github.com/theforeman/smart-proxy/pull/472

I actually like that this does not depend on `wget` anymore. But I agree, it's a bit overengineered. Maybe we can create something similar using [down](https://github.com/janko/down).

lzap replied:
Good point, however my thinking is - if it's a repo that in our current Katello flow:

* Upstream (repo, CDN, custom repo)
* Foreman Server (with Katello)
* Remote Smart Proxy

This actually requires Foreman Server to have access to upstream repos. It should not big deal implementing a proxy-proxy but this creates another complexity and I wonder if anyone wants this. Maybe let's create such an endpoint he moment user asks for it?

timogoebel replied:
We at dmTECH actually have such a use case, that's why I brought it up. In addition, we have to solve this for other (non-RHEL use cases as well).
If Foreman needs access to all media directly, I'd consider this a bug.

Why can't we simply implement a freshness check in the smart-proxy? This seems a lot less intrusive than this change.

lzap commented:
Rebased with new version, tests will still fail but couple of changes:

* Moved the `pxe_files` to host common which enables memorizing of values - in a single request only a pair of HTTP request is made.
* Instead concatenating last-modified, etag and URL the unique ID is calculated in a more clever way - use only last modifier or etag or URL. Therefore resyncing or promoting content actually keeps the unique ID across changes therefore not many filenames are generated. This will prevent https://bugzilla.redhat.com/show_bug.cgi?id=1705979
* Unique ID is now 8 characters short and starts with zero not one (this was a bug which hit me when I was comparing expected results with actual ones).
* Added proper handling of HEAD HTTP redirects which is very common for Pulp repositories - this now fully works.

This version will actually require katello not to override unique id and keep the one from core. Promoting or publishing will have no effect on filename. This is much cleaner and effectively no tftpboot directory cron cleaning will be needed as most (if not all) repository HTTP servers provide last modified headers from what I can tell.

lzap commented:
After some discussion with Timo on IRC, I want to pause and rethink how we got here.

Initially, Foreman PXE naming contention was Osname-Version-Architecture e.g. `Redhat-7.2-x86_64-vmlinuz` or something similar. Katello content is associated to Operating system however extra information like Variant (Server, Workstation) is stored along. When users were flipping these settings, `wget -c` was overwriting PXE files leading to nasty errors.

So we implemented unique ID which is calculated from `URL` and this is now added as a suffix. This was to prevent when user changed media URL to something else without changing OS name. Katello overrides this behavior and puts content view version instead. Now we have reports from the field (Satellite 6.5 beta) that too many unique files are being created in the TFTP directory. We expected this, but the rate at which some customers promote content is high leading to dozens of files *per day*.

In this patch, I want to investigate relying more on content (lastmodified or etag or both) rather than on URL and unify the naming approach for both Foreman core and Katello. That's why I came up with this PR - HTTP HEAD is like a good idea to do because PXE files download is asynchronous so we want to do this before this process starts because it can break (404, 401) and troubleshooting can be time consuming.

Timo made a good point that the HTTP HEAD request should be done on the proxy rather than in Foreman core. We should be able to implement such a HTTP HEAD call in Smart Proxy without breaking its API, the only issue is storing etag or last modified somewhere (probably some kind of timestamp file). The major drawback is when the URL is incorrect, error comes late but Foreman still can roll it back. Orchestration rollback is still better than failed provisioning I think.

So here is my thinking:

* Perform HTTP HEAD on the proxy, raise early if there is 404 or anything wrong and use last-modified/etag headers to initiate re-downloading with `wget`
* Keep using unique ID for core, change Katello to calculate unique ID from OS variant (Server, workstation) or URL instead of from Content View / Version ID which changes way too often.
* Do not care about content change in Foreman core - smart proxy should be responsible for that.
* Close this PR.

timogoebel commented:
I believe this is the correct way to fix this issue.
I do want to get rid of `wget` eventually. But this is definitely out of scope.

We could add a dedicated proxy endpoint that checks if the files are there and implement some more pre-flight checks in Foreman. But that's out of the scope of this PR as well.

lzap commented:
Here is how we could implement this:

* we can keep the current unique id approach in core (in Katello lets change from DBID to URL hash like in core)
* new code in proxy
  * before anything is downloaded, do HTTP HEAD and acquire last-modified/etag sha1 hash
  * then download the content to a filename `SHA1` (or skip if the file already exists)
  * then create a relative symlink `FilenameFromForemanCore-XYZ -> SHA`
  * extend TFTP API with "expiration" flag - every created symlink would have expiration time, this would be stored as `FILENAME.expire_at` timestamp or text file
  * expiration is set by Foreman core according to Token expiration (when it expires TFTP files are actually also invalid) - when set to zero no expiration is performed
* there will be a cronjob running every day cleaning up expired symlinks
* the cronjob can optionally delete `SHA1` named (content) files which are "orphans" (have zero symlinks) and are older than 1 year

We can also use hardlinks instead of symlinks.

timogoebel commented:
> * there will be a cronjob running every day cleaning up expired symlinks

We can just use a background task as we do [in other places of smart-proxy](https://github.com/theforeman/smart-proxy/blob/develop/modules/dhcp_common/free_ips.rb). This way we don't need an actual cronjob (and have a better user experience).
